lib/AST/ASTContext.cpp

```diff
@@ -571,6 +574,38 @@ FuncDecl *ASTContext::getPlusFunctionOnRangeReplaceableCollection() const {
   return Impl.PlusFunctionOnRangeReplaceableCollection;
 }
 
+FuncDecl *ASTContext::getPlusFunctionOnString() const {
+  if (Impl.PlusFunctionOnString) {
+    return Impl.PlusFunctionOnString;
+  }
+  // Find all of the declarations with this name in the Swift module.
+  SmallVector<ValueDecl *, 1> Results;
+  lookupInSwiftModule("+", Results);
+  for (auto Result : Results) {
+    if (auto *FD = dyn_cast<FuncDecl>(Result)) {
+      if (!FD->getOperatorDecl())
+        continue;
+      auto ResultType = FD->getResultInterfaceType();
+      if (ResultType->getNominalOrBoundGenericNominal() != getStringDecl())
+        continue;
+      auto ParamLists = FD->getParameterLists();
+      if (ParamLists.size() != 2 || ParamLists[1]->size() != 2)
+        continue;
+      auto FirstParam = ParamLists[1]->get(0);
+      auto SecondParam = ParamLists[1]->get(1);
+      auto FirstParamType =
+      FirstParam->getInterfaceType()->getNominalOrBoundGenericNominal();
+      auto SecondParamType =
+      SecondParam->getInterfaceType()->getNominalOrBoundGenericNominal();
+      auto StringDecl = getStringDecl();
+      if (FirstParamType == StringDecl && SecondParamType == StringDecl) {
```

Comment:
Nitpick: Can we simplify the above logic by introducing a local closure `isStringParam` and use the closure on both `ParamLsts[1]->get[0]` and `ParamLsts[1]->get[1]`?